File: app/src/main/java/com/pluu/support/tstore/TStoreDetailApi.java
package com.pluu.support.tstore;

import java.util.ArrayList;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import com.pluu.support.impl.AbstractDetailApi;
import com.pluu.webtoon.item.Detail;
import com.pluu.webtoon.item.DetailView;
import com.pluu.webtoon.item.Episode;
import com.pluu.webtoon.item.ShareItem;
import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;

/**
 * TStore 웹툰 상세 API
 * Created by PLUUSYSTEM-NEW on 2015-10-31.
 */
public class TStoreDetailApi extends AbstractDetailApi {

	private static final String MAIN_HOST_URL = "http://m.tstore.co.kr";
	private static final String DETAIL_URL = "http://m.tstore.co.kr/mobilepoc/webtoon/webtoonDetail.omp?prodId=";
	private static final Pattern EPISODE_ID = Pattern.compile("(?<=prodId=)\\w+");
	private static final String SHARE_URL = "http://m.tstore.co.kr/mobilepoc/webtoon/webtoonDetail.omp?prodId=%s&PrePageNm=/detail/webtoon/mw";

	private String id;

	@Override
	public Detail parseDetail(Episode episode) {
		this.id = episode.getEpisodeId();

		Detail ret = new Detail();
		ret.webtoonId = episode.getToonId();
		String response;
		try {
			response = requestApi();
		} catch (Exception e) {
			e.printStackTrace();
			ret.list = new ArrayList<>();
			return ret;
		}

		Document doc = Jsoup.parse(response);
		ret.title = doc.select(".episode-num").text();
		ret.episodeId = id;

		Matcher matcher;
		Elements navi = doc.select(".prev-next a");

		Element temp = navi.first();
		if (temp.hasAttr("href")) {
			matcher = EPISODE_ID.matcher(temp.attr("href"));
			if (matcher.find()) {
				ret.prevLink = matcher.group();
			}
		}

		temp = navi.last();
		if (temp.hasAttr("href")) {
			matcher = EPISODE_ID.matcher(temp.attr("href"));
			if (matcher.find()) {
				ret.nextLink = matcher.group();
			}
		}

		List<DetailView> list = new ArrayList<>();
		Elements scripts = doc.select("script[type=text/javascript]");
		String html;
		Pattern urlPattern = Pattern.compile("(?<=FILE_POS = \\\").+(?=\\\";)");
		Pattern endPattern = Pattern.compile("(?<=bookPages = Number\\(\\\")\\d+(?=\\\"\\))");

		for (Element script : scripts) {
			html = script.html();
			if (html.contains("bookPages")) {
				html = html.replace("\r\n", "");

				Matcher urlMatcher = urlPattern.matcher(html);
				Matcher endMatcher = endPattern.matcher(html);

				if (!urlMatcher.find() || !endMatcher.find()) {
					continue;
				}

				String imgUrl = urlMatcher.group();
				int end = Integer.valueOf(endMatcher.group());

				for (int i = 1; i <= end; i++) {
					list.add(DetailView.createImage(MAIN_HOST_URL + imgUrl + "/" + i + ".jpg"));
				}
				break;
			}
		}

		ret.list = list;
		return ret;
	}

	@Override
	public ShareItem getDetailShare(Episode episode, Detail detail) {
		String url = String.format(SHARE_URL, detail.episodeId);
		ShareItem item = new ShareItem();
		item.title = episode.getTitle() + " / " + detail.title;
		item.url = url;
		return item;
	}

	@Override
	public String getMethod() {
		return GET;
	}

	@Override
	public String getUrl() {
		return DETAIL_URL + id;
	}
}

File: app/src/main/java/com/pluu/support/tstore/TStoreDetailApi.kt
package com.pluu.support.tstore

import android.content.Context
import com.pluu.kotlin.iterator
import com.pluu.support.impl.AbstractDetailApi
import com.pluu.support.impl.NetworkSupportApi
import com.pluu.webtoon.item.Detail
import com.pluu.webtoon.item.DetailView
import com.pluu.webtoon.item.Episode
import com.pluu.webtoon.item.ShareItem
import okhttp3.FormBody
import okhttp3.Request
import org.json.JSONObject
import org.jsoup.Jsoup

/**
 * TStore 웹툰 상세 API
 * Created by pluu on 2017-04-27.
 */
class TStoreDetailApi(context: Context) : AbstractDetailApi(context) {

    private lateinit var id: String

    override fun parseDetail(episode: Episode): Detail {
        this.id = episode.episodeId

        val ret = Detail().apply {
            webtoonId = episode.toonId
            episodeId = id
        }

        val doc = try {
            Jsoup.parse(requestApi())
        } catch (e: Exception) {
            e.printStackTrace()
            ret.list = emptyList<DetailView>()
            return ret
        }

        ret.title = doc.select(".layout-detail-view-header .layout-detail-view-ti")?.text()

        ret.prevLink = doc.select("a[class=btn-episode-link btn-episode-prev]")?.attr("href")?.let {
            EPISODE_ID.find(it)?.value
        }
        ret.nextLink = doc.select("a[class=btn-episode-link btn-episode-next]")?.attr("href")?.let {
            EPISODE_ID.find(it)?.value
        }
        doc.select("#ifrmResizing")?.attr("src")?.let {
            CURRENT_ID.find(it)?.value?.apply {
                ret.list = getList(this)
            }
        }
        return ret
    }

    private fun getList(id: String): List<DetailView> {
        val request = Request.Builder().apply {
            url("https://www.myktoon.com/api/work/getTimesDetailImageList.kt")
            // POST
            val requestBody = FormBody.Builder().apply {
                add("timesseq", id)
            }.build()
            addHeader("Referer", "https://www.myktoon.com/web/open/build/work_view.kt")
            post(requestBody)
        }

        val list = mutableListOf<DetailView>()
        JSONObject(requestApi(request.build())).optJSONArray("imageList")?.iterator()?.forEach {
            list.add(DetailView.createImage(it.optString("imagepath")).apply {
                height = it.optInt("height").toFloat()
            })
        }
        return list
    }

    override fun getDetailShare(episode: Episode, detail: Detail) = ShareItem(
        title = "${episode.title} / ${detail.title}",
        url = "http://m.tstore.co.kr/mobilepoc/webtoon/webtoonDetail.omp?prodId=${detail.episodeId}&PrePageNm=/detail/webtoon/mw"
    )

    override val url: String
        get() = "http://m.onestore.co.kr/mobilepoc/webtoon/webtoonDetail.omp?prodId=$id&PrePageNm="

    override val method: String
        get() = NetworkSupportApi.GET

    companion object {
        private val EPISODE_ID = "(?<=ajaxWebtoonDetail\\(\\\\')[A-Za-z0-9]+".toRegex()
        private val CURRENT_ID = "(?<=timesseq=)\\d+".toRegex()
    }
}
